pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java

```diff
@@ -79,28 +80,23 @@ private void attachTag(FlowNode currentNode, String syntheticContext) {
         }
     }
 
-    public static class GraphListenerAction extends InvisibleAction implements RunAction2 {
+    @Extension
+    public static class FlowExecutionListenerImpl extends FlowExecutionListener {
         @Override
-        public void onLoad(Run<?, ?> r) {
-            if (r != null && r instanceof WorkflowRun) {
-                WorkflowRun run = (WorkflowRun) r;
-                attachListener(run);
-            }
-        }
-
-        @Override
-        public void onAttached(Run<?, ?> r) {
-            if (r != null && r instanceof WorkflowRun) {
-                WorkflowRun run = (WorkflowRun) r;
-                attachListener(run);
-            }
-        }
-
-        private void attachListener(WorkflowRun run) {
-            if (run != null) {
-                FlowExecution exec = run.getExecution();
-                if (exec != null && !exec.isComplete()) {
-                    exec.addListener(new SyntheticStageGraphListener());
+        public void onRunning(FlowExecution execution) {
+            if (execution != null && !execution.isComplete()) {
+                // Only attach the graph listener if we know the run this execution is for is a Declarative run,
+                // i.e., if it has an ExecutionModelAction.
+                try {
+                    FlowExecutionOwner owner = execution.getOwner();
+                    if (owner != null && owner.getExecutable() instanceof WorkflowRun) {
+                        WorkflowRun r = (WorkflowRun)owner.getExecutable();
+                        if (r.getAction(ExecutionModelAction.class) != null) {
+                            execution.addListener(new SyntheticStageGraphListener());
```

Comment:
Yes you would have to follow the same chain of APIs to determine if this `FlowNode` is from a Declarative build. What _is_ simpler is that you are just reacting passively to events. There is no listener to attach, no risk that you have the timing wrong and there is some kind of subtle race condition, no risk of attaching the same listener twice, etc. So the patch would be shorter (and IMO clearer and safer) than what you have now.